Spigot-Server-Patches/0236-Implemented-the-cancellable-ban-events.patch

```diff
@@ -0,0 +1,271 @@
+         return astring;
+     }
+ 
++    // Paper start
++    @Override
++    public void add(GameProfileBanEntry entry){
++        MinecraftServer server = MinecraftServer.getServer();
++        GameProfile profile = entry.getKey();
++
++        BanProfileEvent event = CraftEventFactory.callEvent(new BanProfileEvent(new CraftProfileBanEntry(profile, entry, server.getPlayerList().getProfileBans() )));
++        if(event.isCancelled()) return;
++
++        entry.setCreated(event.getCreated());
++        entry.setExpiration(event.getExpiration());
++        entry.setSource(event.getSource());
++        entry.setReason(event.getReason());
++        super.add(entry);
++    }
++    // Paper end
++
+     protected String b(GameProfile gameprofile) {
+         return gameprofile.getId().toString();
+     }
+diff --git a/src/main/java/net/minecraft/server/IpBanList.java b/src/main/java/net/minecraft/server/IpBanList.java
+index 617d326f..1368c715 100644
+--- a/src/main/java/net/minecraft/server/IpBanList.java
++++ b/src/main/java/net/minecraft/server/IpBanList.java
+@@ -1,6 +1,9 @@
+ package net.minecraft.server;
+ 
+ import com.google.gson.JsonObject;
++import org.bukkit.craftbukkit.CraftIpBanEntry;
++import org.bukkit.event.server.BanIpEvent;
++
+ import java.io.File;
+ import java.net.SocketAddress;
+ 
+@@ -26,6 +29,22 @@ public class IpBanList extends JsonList<String, IpBanEntry> {
+         return (IpBanEntry) this.get((Object) s);
+     }
+ 
++    // Paper start
++    @Override public void add(IpBanEntry entry){
++        MinecraftServer server = MinecraftServer.getServer();
++        String target = entry.getKey();
++
++        BanIpEvent event= new BanIpEvent(new CraftIpBanEntry(target, entry, server.getPlayerList().getIPBans()));
++        if(event.isCancelled()) return;
++
++        entry.setCreated(event.getCreated()); // repeated code, override in superclass instead? (ExpirableListEntry)
++        entry.setExpiration(event.getExpiration());
++        entry.setSource(event.getSource());
++        entry.setReason(event.getReason());
++        super.add(entry);
++    }
++    // Paper end
++
+     private String c(SocketAddress socketaddress) {
+         String s = socketaddress.toString();
+ 
+diff --git a/src/main/java/net/minecraft/server/JsonList.java b/src/main/java/net/minecraft/server/JsonList.java
+index 0859c7eb..49d5b42d 100644
+--- a/src/main/java/net/minecraft/server/JsonList.java
++++ b/src/main/java/net/minecraft/server/JsonList.java
+@@ -82,7 +82,9 @@ public class JsonList<K, V extends JsonListEntry<K>> {
+ 
+     }
+ 
+-    public V get(K k0) {
++    // Paper start
++    public V get(Object k0) {
++        // Paper end
+         this.h();
+         return (V) this.d.get(this.a(k0)); // CraftBukkit - fix decompile error
+     }
+@@ -112,7 +114,8 @@ public class JsonList<K, V extends JsonListEntry<K>> {
+         return this.d.size() < 1;
+     }
+ 
+-    protected String a(K k0) {
++    //Paper Start - fix decompile error
```

Comment:
This doesn't have a Paper end, but neither this nor the get() method even need to use the start/end style. Just do a single line change/comment.